### PR TITLE
feat: Implement str predicate function contract

### PR DESCRIPTION
**Pull request by HikiShi**

### CODE REVIEW

File: dna/src/commonMain/kotlin/predicate/PredicateFunctionContract.kt [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package com.github.trilobiitti.trilobite.dna.predicate
+
+import com.github.trilobiitti.trilobite.dna.data.DataType
+import kotlin.reflect.KClass
+
+
+interface PredicateFunctionContract {
+    // Throws if argument types are not acceptable
+    // Returns result type if they are
```
Comment (AlexeyBond) [inline dna/src/commonMain/kotlin/predicate/PredicateFunctionContract.kt]: This should be formatted as [JavaDoc](https://en.wikipedia.org/wiki/Javadoc)

File: dna/src/commonMain/kotlin/predicate/PredicateFunctionContract.kt [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    fun validateArgumentTypes(inputs: List<PredicateFunctionInput>): DataType<*>
+}
+
+
+class NotValidPredicateFunctionInputError(val availableFunctionInputClasses: List<KClass<*>> )
+    : Error("Not valid predicate function input. Allowed classes:" +
+        " ${availableFunctionInputClasses.map { it.simpleName }.joinToString(separator = " ,") }}." +
+        "Combination of these classes is not allowed")
```
Comment (AlexeyBond) [inline dna/src/commonMain/kotlin/predicate/PredicateFunctionContract.kt]: There is no rule disallowing combinations of literals and non-literals as inputs of a predicate function in general case. Also this exception class doesn't cover all possible scenarios of inputs invalidity, so let's just reuse `ValidationError` here instead of creating a new error class for now.

File: dna/src/commonMain/kotlin/predicate/PredicateFunctionContractStrImpl.kt [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+
+    private fun literalStrategy(inputs:  List<PredicateFunctionInput>): Nothing = TODO();
+    private fun valueStrategy(inputs:  List<PredicateFunctionInput>): Nothing = TODO();
+
+
+    override fun validateArgumentTypes(inputs: List<PredicateFunctionInput>): DataType<*>  = when {
+        isUseLiteralStrategy(inputs) ->  literalStrategy(inputs)
+        isUseValueStrategy(inputs) ->  valueStrategy(inputs)
+        else -> throw NotValidPredicateFunctionInputError(availableFunctionInputClasses = listOf(LiteralInput::class, ValueInput::class ))
```
Comment (AlexeyBond) [inline dna/src/commonMain/kotlin/predicate/PredicateFunctionContractStrImpl.kt]: Since `PredicateFunctionInput` is sealed, there can be no other subclasses except `LiteralInput` and `ValueInput`. Also "str" shouldn't accept anything but a single literal argument, so this kinda doesn't make sense

File: dna/src/commonTest/kotlin/predicate/PredicateFunctionContractTest.kt [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        assertFailsWith(NotValidPredicateFunctionInputError::class) {
+            predicateStrFunctionContract.validateArgumentTypes(listOf(LiteralInput("updateAt"), ValueInput(NullableType(DataTypes.NUMBER)) ))
+        }
+    }
+
+    @Test
+    fun shouldReturnNullableStringDataTypeWhenPassValidPredicateFunctionInput() {
+        val dataType = predicateStrFunctionContract.validateArgumentTypes(listOf(LiteralInput("name")))
+
+        assertEquals(NullableType(DataTypes.STRING), dataType)
```
Comment (AlexeyBond) [inline dna/src/commonTest/kotlin/predicate/PredicateFunctionContractTest.kt]: "str" always returns a string, so output type should not be nullable

File: dna/src/commonMain/kotlin/predicate/PredicateFunctionContractStrImpl.kt [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    }
+}
```
Comment (AlexeyBond) [inline dna/src/commonMain/kotlin/predicate/PredicateFunctionContractStrImpl.kt]: newlines!

File: dna/src/commonMain/kotlin/predicate/PredicateFunctionContractStrImpl.kt [hunk 6]
```diff
@@ -0,0 +1,20 @@
+package com.github.trilobiitti.trilobite.dna.predicate
+
+import com.github.trilobiitti.trilobite.dna.data.DataType
+
+class PredicateFunctionContractStrImpl: PredicateFunctionContract {
```
Comment (AlexeyBond) [inline dna/src/commonMain/kotlin/predicate/PredicateFunctionContractStrImpl.kt]: This can be a singleton object
```kotlin
object StrFunctionContract: PredicateFunctionContract { ... }
```

Comment (AlexeyBond) [inline dna/src/commonMain/kotlin/predicate/PredicateFunctionContractStrImpl.kt]: Also, please enable "reformat code" option in precommit actions settings (or reformat it manually)

Comment (AlexeyBond) [inline dna/src/commonMain/kotlin/predicate/PredicateFunctionContractStrImpl.kt]: Also, let's group all contracts into a single module (e.g. `StandardContracts.kt`) and make `predicate` package a subpackage of `data` package

**Comment by AlexeyBond** [comment]
We should create a feature-branch for first iteration of this predicatish story

File: dna/src/commonTest/kotlin/predicate/PredicateFunctionContractTest.kt [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+import kotlin.test.assertFailsWith
+
+
+class PredicateStrFunctionContractTest {
+
+    private val predicateStrFunctionContract = PredicateFunctionContractStrImpl();
+
+
+    @Test
+    fun shouldThrowsWhenPassNotValidPredicateFunctionInput(){
```
Comment (AlexeyBond) [inline dna/src/commonTest/kotlin/predicate/PredicateFunctionContractTest.kt]:
```suggestion
    fun shouldThrowWhenInvalidPredicateFunctionInputsArePassed(){
```

also, this test doesn't cover all edge cases

File: dna/src/commonMain/kotlin/data/predicate/StandardContracts.kt [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+object StrFunctionContract : PredicateFunctionContract {
+    private fun inputSizeValidation(inputs: List<PredicateFunctionInput>) {
+        if (inputs.isEmpty() || inputs.size > 1) {
+            throw SimpleValidationError("One parameter must be passed");
+        }
+    }
+
+    private fun inputTypeValidation(inputs: List<PredicateFunctionInput>) {
+        if (!inputs.all { it is LiteralInput }) {
+            throw SimpleValidationError("Only ${LiteralInput::class.simpleName}  are accepted");
```
Comment (AlexeyBond) [inline dna/src/commonMain/kotlin/data/predicate/StandardContracts.kt]: Those, who will see this message may be unaware of validator's implementation details. Error message should be provided in terms of predicate language

File: dna/src/commonMain/kotlin/data/predicate/StandardContracts.kt [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+object StrFunctionContract : PredicateFunctionContract {
+    private fun inputSizeValidation(inputs: List<PredicateFunctionInput>) {
+        if (inputs.isEmpty() || inputs.size > 1) {
+            throw SimpleValidationError("One parameter must be passed");
+        }
+    }
+
+    private fun inputTypeValidation(inputs: List<PredicateFunctionInput>) {
+        if (!inputs.all { it is LiteralInput }) {
```
Comment (AlexeyBond) [inline dna/src/commonMain/kotlin/data/predicate/StandardContracts.kt]: We can use `inputs.first` as we already know that there is one and only one element.

File: dna/src/commonMain/kotlin/data/predicate/StandardContracts.kt [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+            throw SimpleValidationError("Only ${LiteralInput::class.simpleName}  are accepted");
+        }
+    }
+
+    override fun validateArgumentTypes(inputs: List<PredicateFunctionInput>): DataType<*> {
+        inputSizeValidation(inputs);
+        inputTypeValidation(inputs);
```
Comment (AlexeyBond) [inline dna/src/commonMain/kotlin/data/predicate/StandardContracts.kt]: I don't see much sense in separate methods here. Conditions and error messages are self-explaining enough

**Review by AlexeyBond — APPROVED** [review]
